File: thrifty-schema/src/main/java/com/microsoft/thrifty/schema/Schema.java
/*
 * Thrifty
 *
 * Copyright (c) Microsoft Corporation
 *
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING
 * WITHOUT LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE,
 * FITNESS FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
 *
 * See the Apache Version 2.0 License for specific language governing permissions and limitations under the License.
 */
package com.microsoft.thrifty.schema;

import com.google.common.collect.ImmutableList;

import java.util.NoSuchElementException;

/**
 * Encapsulates all types, values, and services defined in a set of Thrift
 * files.
 *
 * Strictly speaking, this is a lossy representation - the original filesystem
 * structure of the source Programs is not preserved.  As such, this isn't a
 * suitable representation from which to generate C++ code, or any other
 * module-based language for the matter.  But as we're only concerned with Java
 * here, it's perfectly convenient.
 */
public class Schema {
    private final ImmutableList<StructType> structs;
    private final ImmutableList<StructType> unions;
    private final ImmutableList<StructType> exceptions;
    private final ImmutableList<EnumType> enums;
    private final ImmutableList<Constant> constants;
    private final ImmutableList<Typedef> typedefs;
    private final ImmutableList<Service> services;

    Schema(Iterable<Program> programs) {
        ImmutableList.Builder<StructType> structs    = ImmutableList.builder();
        ImmutableList.Builder<StructType> unions     = ImmutableList.builder();
        ImmutableList.Builder<StructType> exceptions = ImmutableList.builder();
        ImmutableList.Builder<EnumType>   enums      = ImmutableList.builder();
        ImmutableList.Builder<Constant>   constants  = ImmutableList.builder();
        ImmutableList.Builder<Typedef>    typedefs   = ImmutableList.builder();
        ImmutableList.Builder<Service>    services   = ImmutableList.builder();

        for (Program program : programs) {
            structs.addAll(program.structs());
            unions.addAll(program.unions());
            exceptions.addAll(program.exceptions());
            enums.addAll(program.enums());
            constants.addAll(program.constants());
            typedefs.addAll(program.typedefs());
            services.addAll(program.services());
        }

        this.structs = structs.build();
        this.unions = unions.build();
        this.exceptions = exceptions.build();
        this.enums = enums.build();
        this.constants = constants.build();
        this.typedefs = typedefs.build();
        this.services = services.build();
    }

    private Schema(Builder builder) {
        this.structs = builder.structs;
        this.unions = builder.unions;
        this.exceptions = builder.exceptions;
        this.enums = builder.enums;
        this.constants = builder.constants;
        this.typedefs = builder.typedefs;
        this.services = builder.services;
    }

    public ImmutableList<StructType> structs() {
        return structs;
    }

    public ImmutableList<StructType> unions() {
        return unions;
    }

    public ImmutableList<StructType> exceptions() {
        return exceptions;
    }

    public ImmutableList<EnumType> enums() {
        return enums;
    }

    public ImmutableList<Constant> constants() {
        return constants;
    }

    public ImmutableList<Typedef> typedefs() {
        return typedefs;
    }

    public ImmutableList<Service> services() {
        return services;
    }

    public EnumType findEnumByType(ThriftType type) {
        for (EnumType enumType : enums) {
            if (enumType.type().equals(type)) {
                return enumType;
            }
        }
        throw new NoSuchElementException("No enum type matching " + type.name());
    }

    public Builder toBuilder() {
        return new Builder(structs, unions, exceptions, enums, constants, typedefs, services);
    }

    public static final class Builder {
        private ImmutableList<StructType> structs;
        private ImmutableList<StructType> unions;
        private ImmutableList<StructType> exceptions;
        private ImmutableList<EnumType> enums;
        private ImmutableList<Constant> constants;
        private ImmutableList<Typedef> typedefs;
        private ImmutableList<Service> services;

        Builder(ImmutableList<StructType> structs,
                ImmutableList<StructType> unions,
                ImmutableList<StructType> exceptions,
                ImmutableList<EnumType> enums,
                ImmutableList<Constant> constants,
                ImmutableList<Typedef> typedefs,
                ImmutableList<Service> services) {
            this.structs = structs;
            this.unions = unions;
            this.exceptions = exceptions;
            this.enums = enums;
            this.constants = constants;
            this.typedefs = typedefs;
            this.services = services;
        }

        public Builder structs(ImmutableList<StructType> structs) {
            if (structs == null) {
                throw new NullPointerException("structs may not be null");
            }
            this.structs = structs;
            return this;
        }

        public Builder unions(ImmutableList<StructType> unions) {
            if (structs == null) {
                throw new NullPointerException("unions may not be null");
            }
            this.unions = unions;
            return this;
        }

        public Builder exceptions(ImmutableList<StructType> exceptions) {
            if (structs == null) {
                throw new NullPointerException("exceptions may not be null");
            }
            this.exceptions = exceptions;
            return this;
        }

        public Builder enums(ImmutableList<EnumType> enums) {
            if (structs == null) {
                throw new NullPointerException("enums may not be null");
            }
            this.enums = enums;
            return this;
        }

        public Builder constants(ImmutableList<Constant> constants) {
            if (structs == null) {
                throw new NullPointerException("constants may not be null");
            }
            this.constants = constants;
            return this;
        }

        public Builder typedefs(ImmutableList<Typedef> typedefs) {
            if (structs == null) {
                throw new NullPointerException("typedefs may not be null");
            }
            this.typedefs = typedefs;
            return this;
        }

        public Builder services(ImmutableList<Service> services) {
            if (structs == null) {
                throw new NullPointerException("services may not be null");
            }
            this.services = services;
            return this;
        }

        public Schema build() {
            return new Schema(this);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) { return true; }
        if (o == null || getClass() != o.getClass()) { return false; }

        Schema schema = (Schema) o;

        if (!structs.equals(schema.structs)) { return false; }
        if (!unions.equals(schema.unions)) { return false; }
        if (!exceptions.equals(schema.exceptions)) { return false; }
        if (!enums.equals(schema.enums)) { return false; }
        if (!constants.equals(schema.constants)) { return false; }
        if (!typedefs.equals(schema.typedefs)) { return false; }
        return services.equals(schema.services);

    }

    @Override
    public int hashCode() {
        int result = structs.hashCode();
        result = 31 * result + unions.hashCode();
        result = 31 * result + exceptions.hashCode();
        result = 31 * result + enums.hashCode();
        result = 31 * result + constants.hashCode();
        result = 31 * result + typedefs.hashCode();
        result = 31 * result + services.hashCode();
        return result;
    }
}

File: thrifty-schema/src/main/kotlin/com/microsoft/thrifty/schema/Schema.kt
/*
 * Thrifty
 *
 * Copyright (c) Microsoft Corporation
 *
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING
 * WITHOUT LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE,
 * FITNESS FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
 *
 * See the Apache Version 2.0 License for specific language governing permissions and limitations under the License.
 */
package com.microsoft.thrifty.schema

/**
 * Encapsulates all types, values, and services defined in a set of Thrift
 * files.
 *
 * Strictly speaking, this is a lossy representation - the original filesystem
 * structure of the source Programs is not preserved.  As such, this isn't a
 * suitable representation from which to generate C++ code, or any other
 * module-based language for the matter.  But as we're only concerned with Java
 * here, it's perfectly convenient.
 */
class Schema {
    /**
     * All `struct` entities contained in the parsed .thrift files
     */
    val structs: List<StructType>

    /**
     * All `union` entities contained in the parsed .thrift files
     */
    val unions: List<StructType>

    /**
     * All `exception` entities contained in the parsed .thrift files
     */
    val exceptions: List<StructType>

    /**
     * All `enum` types defined in the parsed .thrift files.
     */
    val enums: List<EnumType>

    /**
     * All `const` elements defined in the parsed .thrift files.
     */
    val constants: List<Constant>

    /**
     * All `typedef` aliases defined in the parsed .thrift files.
     */
    val typedefs: List<TypedefType>

    /**
     * All `service` types defined in the parsed .thrift files.
     */
    val services: List<ServiceType>

    /**
     * @return an [Iterable] of all [UserElements][UserElement] in this [Schema].
     */
    fun elements(): Iterable<UserElement> {
        return structs + unions + exceptions + enums + constants + typedefs + services
    }

    internal constructor(programs: Iterable<Program>) {
        val structs = mutableListOf<StructType>()
        val unions = mutableListOf<StructType>()
        val exceptions = mutableListOf<StructType>()
        val enums = mutableListOf<EnumType>()
        val constants = mutableListOf<Constant>()
        val typedefs = mutableListOf<TypedefType>()
        val services = mutableListOf<ServiceType>()

        for (program in programs) {
            structs.addAll(program.structs)
            unions.addAll(program.unions)
            exceptions.addAll(program.exceptions)
            enums.addAll(program.enums)
            constants.addAll(program.constants)
            typedefs.addAll(program.typedefs)
            services.addAll(program.services)
        }

        this.structs = structs
        this.unions = unions
        this.exceptions = exceptions
        this.enums = enums
        this.constants = constants
        this.typedefs = typedefs
        this.services = services
    }

    private constructor(builder: Builder) {
        this.structs = builder.structs
        this.unions = builder.unions
        this.exceptions = builder.exceptions
        this.enums = builder.enums
        this.constants = builder.constants
        this.typedefs = builder.typedefs
        this.services = builder.services
    }

    /**
     * Returns a [Builder] initialized with this schema's types.
     */
    fun toBuilder(): Builder {
        return Builder(structs, unions, exceptions, enums, constants, typedefs, services)
    }

    /**
     * A builder for [schemas][Schema].
     */
    class Builder internal constructor(
            internal var structs: List<StructType>,
            internal var unions: List<StructType>,
            internal var exceptions: List<StructType>,
            internal var enums: List<EnumType>,
            internal var constants: List<Constant>,
            internal var typedefs: List<TypedefType>,
            internal var services: List<ServiceType>
    ) {
        /**
         * Use the given [structs] for the schema under construction.
         */
        fun structs(structs: List<StructType>): Builder = apply {
            this.structs = structs.toList()
        }

        /**
         * Use the given [unions] for the schema under construction.
         */
        fun unions(unions: List<StructType>): Builder = apply {
            this.unions = unions.toList()
        }

        /**
         * Use the given [exceptions] for the schema under construction.
         */
        fun exceptions(exceptions: List<StructType>): Builder = apply {
            this.exceptions = exceptions.toList()
        }

        /**
         * Use the given [enums] for the schema under construction.
         */
        fun enums(enums: List<EnumType>): Builder = apply {
            this.enums = enums.toList()
        }

        /**
         * Use the given [constants] for the schema under construction.
         */
        fun constants(constants: List<Constant>): Builder = apply {
            this.constants = constants.toList()
        }

        /**
         * Use the given [typedefs] for the schema under construction.
         */
        fun typedefs(typedefs: List<TypedefType>): Builder = apply {
            this.typedefs = typedefs.toList()
        }

        /**
         * Use the given [services] for the schema under construction.
         */
        fun services(services: List<ServiceType>): Builder = apply {
            this.services = services.toList()
        }

        /**
         * Build a new [Schema].
         */
        fun build(): Schema = Schema(this)
    }

    /** @inheritdoc */
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Schema

        if (structs != other.structs) return false
        if (unions != other.unions) return false
        if (exceptions != other.exceptions) return false
        if (enums != other.enums) return false
        if (constants != other.constants) return false
        if (typedefs != other.typedefs) return false
        if (services != other.services) return false

        return true
    }

    /** @inheritdoc */
    override fun hashCode(): Int {
        var result = structs.hashCode()
        result = 31 * result + unions.hashCode()
        result = 31 * result + exceptions.hashCode()
        result = 31 * result + enums.hashCode()
        result = 31 * result + constants.hashCode()
        result = 31 * result + typedefs.hashCode()
        result = 31 * result + services.hashCode()
        return result
    }
}
